merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/InsertActivityTypesAction.java

```diff
@@ -0,0 +1,71 @@
+package gov.nasa.jpl.aerie.merlin.server.remotes.postgres;
+
+import gov.nasa.jpl.aerie.merlin.server.models.ActivityType;
+import org.intellij.lang.annotations.Language;
+
+import java.sql.BatchUpdateException;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Collection;
+
+import static gov.nasa.jpl.aerie.merlin.driver.json.ValueSchemaJsonParser.valueSchemaP;
+
+/*package-local*/ final class InsertActivityTypesAction implements AutoCloseable {
+  private static final @Language("SQL") String sql = """
+    insert into activity_type (model_id, name, parameters, required_parameters, computed_attributes_value_schema)
+    values (?, ?, ?::json, ?::json, ?::json)
+    on conflict (model_id, name) do update
+      set parameters = excluded.parameters,
+      required_parameters = excluded.required_parameters,
+      computed_attributes_value_schema = excluded.computed_attributes_value_schema
+    """;
+
+  private final PreparedStatement statement;
+
+  public InsertActivityTypesAction(final Connection connection) throws SQLException {
+    this.statement = connection.prepareStatement(sql);
+  }
+
+  public void apply(final int modelId, Collection<ActivityType> activityTypes)
+  throws SQLException, FailedInsertException
+  {
+    final var connection = statement.getConnection();
+    try {
+      // From the docs (https://docs.oracle.com/javase/tutorial/jdbc/basics/retrieving.html#batch_updates):
+      // "To allow for correct error handling, you should always disable auto-commit mode before beginning a batch update."
+      connection.setAutoCommit(false);
```

Comment:
JFYI - this logic for toggling auto commit was the _raison d'être_ of the TransactionContext.

No change requested - what you have is just fine for this case 👍